entr.1
.Dd October 1, 2026
.Dt ENTR 1
.Os
.Sh NAME
.Nm entr
.Nd run arbitrary commands when files change
.Sh SYNOPSIS
.Nm
.Op Fl acdnprsxz
.Ar utility
.Op Ar argument /_ ...
.Sh DESCRIPTION
A list of files provided on standard input, and the
.Ar utility
is executed using the supplied arguments if any of them change.
.Nm
waits for the child process to finish before responding to subsequent file
system events.
A TTY is also opened before entering the watch loop in order to support
interactive utilities.
.Pp
The arguments are as follows:
.Bl -tag -width Ds
.It Fl a
Respond to all events which occur while the
.Ar utility
is running.
Without this option,
.Nm
consolidates events in order to avoid looping.
This option has no effect in conjunction with the
.Fl r
flag.
.It Fl c
Clear the screen before invoking the
.Ar utility
specified on the command line.
Specify twice to erase the scrollback buffer.
.It Fl d
Track the parent directory of regular files provided as input.
If specified twice, all new entries to a directory are recognized, otherwise
files with names beginning with
.Ql \&.
are ignored.
.It Fl n
Run in non-interactive mode.
In this mode
.Nm entr
does not attempt to read from the TTY or change its properties.
.It Fl p
Postpone the first execution of the
.Ar utility
until a file is modified.
.It Fl r
Reload a persistent child process.
As with the standard mode of operation, a
.Ar utility
which terminates is not executed again until a file system or keyboard event is
processed.
A process group is created to prevent shell scripts from masking signals.
.Nm
waits for the
.Ar utility
to exit to ensure that resources such as sockets have been closed.
Control of the TTY is not transferred to the child process.
.It Fl s
Evaluate the first argument using the interpreter specified by the
.Ev SHELL
environment variable.
.It Fl x
Format custom exit status messages using a persistent
.Xr awk 1
process.
The path to the status script is defined by the environment variable
.Ev ENTR_STATUS_SCRIPT .
If the status script does not exist,
.Nm
will create an example.
Shell commands and file redirection is not permitted by default, but may be
enabled by specifying
.Fl x
twice.
.It Fl z
Exit after the
.Ar utility
completes.
When combined with
.Fl r
the
.Ar utility
will be restarted again only in response to commands or file system
events.
.El
.Pp
The first argument named
.Ar /_
is replaced with the absolute path of the first file to trigger an event.
The first file under watch is used as the default.
If the
.Fl s
option is used, the name of the first file to trigger an event can be read from
.Va $0 .
.Sh COMMANDS
.Nm
listens for keyboard input and responds to the following commands:
.Bl -tag -width <space>
.It Aq Cm space
Execute the utility immediately.
If the
.Fl Cm r
option is set this will terminate and restart the child process as if a file
change event had occurred.
.It Cm q
Quit; equivalent pressing
.Aq Cm control-C .
.El
.Sh ENVIRONMENT
.Bl -tag -width "ENTR_ENVIRON"
.It Ev ENTR_STATUS_SCRIPT
The path to the exit status script enabled by the
.Fl x
option.
By default
.Pa $HOME/.entr/status.awk
is evaluated.
.It Ev ENTR_RESTART_SIGNAL
Signal used to terminate the
.Ar utility .
Supported signal names are
.Dv HUP ,
.Dv INT ,
.Dv QUIT ,
.Dv TERM ,
.Dv USR1
and
.Dv USR2 .
The default is
.Dv SIGTERM .
.It Ev EV_TRACE
Print file system event messages.
.It Ev PAGER
Set to
.Pa /bin/cat
by default
to prevent interactive utilities from waiting for keyboard input if output does
not fit on the screen.
.It Ev SHELL
Specify the shell to use with the
.Fl s
flag.
The default is
.Pa /bin/sh .
.El
.Sh EXIT STATUS
If the
.Fl z
flag is set and the
.Ar utility
is successfully executed, the status of the child process is returned.
If the child process was terminated by a signal, the exit status is the signal
number plus 128.
.Pp
.Nm
normally returns one of the following values:
.Pp
.Bl -tag -width Ds -offset indent -compact
.It 0
Normal termination after receiving
.Dv SIGINT
.It 1
No regular files were provided as input or an error occurred
.It 2
Files were added or removed from a directory
.El
.Sh EXAMPLES
Rebuild a project if source files change, limiting output to the first 20 lines:
.Pp
.Dl $ find src/ | entr -s 'make | head -n 20'
.Pp
Launch and auto-reload a node.js server:
.Pp
.Dl $ ls *.js | entr -r node app.js
.Pp
Clear the screen and run a query after the SQL script is updated:
.Pp
.Dl $ echo my.sql | entr -cp psql -f /_
.Pp
Rebuild project if a source file is modified or added to the src/ directory:
.Pp
.Dl $ while sleep 0.1; do ls -d src src/*.rb | entr make; done
.Pp
Auto-reload a web server, or terminate if the server exits
.Pp
.Dl $ ls * | entr -rz ./httpd
.Sh AUTHORS
.An Eric Radman Aq Mt ericshane@eradman.com


NEWS
= Release History

== 5.9: October 1, 2026

 - Sleep for 1ms after utility is terminated in restart mode
 - Abort if path length exceed boundaries, no longer rely on strlcpy(3)
 - Check argument length before copying

== 5.8: March 2, 2026

 - MacOS: set open_max limit using MAXFILESPERPROC
 - Add usage hint and '-h' flag to display option summary
 - Add return value checks for malloc(3) and pipe(2)
 - Make signal number configurable with ENTR_RESTART_SIGNAL
 - Monitor symlinks on Linux and MacOS
 - Avoid expanding filenames using realpath(3)
 - Linux: return correct status if poll(2) is interrupted
 - Always accept directories as input, making '-d' flag optional

== 5.7: February 6, 2025

 - Update MANPREFIX for FreeBSD
 - Format source using clang-format
 - Reorganize system tests to run without a TTY
 - Remove unused X option character

== 5.6: July 1, 2024

 - Process exit status messages using background awk scripts
 - Use min(OPEN_MAX, rl.rlim_max) to set rlim_cur on MacOS
 - Detect and use native strlcpy(3)
 - Always build with GNU_SOURCE set on Linux

== 5.5: November 17, 2023

 - Report correct error if open(3) fails
 - Use kern.maxfilesperproc to determine file open limit on MacOS

== 5.4: August 1, 2023

 - 'make test' runs a quick smoketest, 'make check' runs regressions
 - Set IN_CLOEXEC only for inotify_init, kqueue uses similar setting by default
 - Unconditionally try to set soft file limit to 2^16 on MacOS
 - Use non-reentrant calls sparingly in signal handlers
 - configure: use TARGET_OS to override the output of uname(1)

== 5.3: January 20, 2023

 - Symlink changes detected on Linux by setting 'ENTR_INOTIFY_SYMLINK'
 - Use /dev/null rather then closed pipe for stdin in -r mode
 - Utilize {O,FD}_CLOEXEC flag for unintentional leaks of descriptors to
   executed utilities
 - Remove C unit tests
 - Only respond to attribute/inode changes on Linux

== 5.2: May 05, 2022

 - Update copy of strlcpy(3) for Linux
 - Detect file deletion from directories on Linux
 - Print the signal that terminated a child when using '-s'
 - Return 128+signal that terminated a child when using '-z'
 - Ensure terminal settings are reset when '-z' is set

== 5.1: December 28, 2021

 - Detect files moved to or from directories on Linux
 - Allow detection of directory entries beginning with '.' by specifying '-d' twice
 - Only reset terminal settings in exit handler if settings were changed

== 5.0: July 2, 2021

 - Add missing call to getrlimit(3) on MacOS
 - Eliminate memory management warnings on Linux
 - EV_TRACE prints file mode and file name
 - Ignore asynchronous attribute changes on MacOS 11

== 4.9: May 3, 2021

 - EV_TRACE also prints file/notify descriptor limit
 - Don't raise rlim_cur on MacOS
 - Set 2^16 watches if inotify limits cannot be read
 - Raise an error and suggest '-n' if terminal attributes cannot be read

== 4.8: February 26, 2021

 - Set a maximum of 2^19 watches to guard against absurd file open limits on
   MacOS
 - Use control sequences to clear the display and specify '-c' twice to erase
   the scrollback buffer

== 4.7: January 29, 2021

 - Use system file descriptor limits when max_user_watches is not accessible
 - Return the exit status of the child process when the '-z' option is used
 - Handle SIGHUP so child process are terminated when a terminal is closed
 - More accurately return shell exit code using '-s' option

== 4.6: July 1, 2020

 - Always call waitpid(2) to avoid dead processes
 - Duplicate STDIN file descriptor before closing; for the '-r' option

== 4.5: April 20, 2020

 - New '-z' "one-shot" option self-terminates after the utility exits
 - Termination by 'q' or 'SIGINT' results in an exit status of 0

== 4.4: January 02, 2020

 - Use a single inotify queue on Linux, limited by
   /proc/sys/fs/inotify/max_user_watches
 - Set the environment variable 'ENTR_INOTIFY_WORKAROUND' to enable a
   compatibility mode for platforms with deformed inotify support

== 4.3: September 25, 2019

 - Move hosting from bitbucket.org to github.com
 - Rename tag names from entr-X.Y to X.Y
 - Replace hg with git as a dependency for system-test.sh

== 4.2: March 28, 2019

 - Use pledge(2) on OpenBSD
 - New '-a' option enables response to events that occur while the utility is
   running
 - Error messages direct users to http://eradman.com/entrproject/limits.html
 - Correctly report error when a file cannot be reopened

== 4.1: June 13, 2018

 - New '-n' non-interactive option disables keyboard input
 - EV_TRACE environment variable enables file system event tracing
 - Track changes to the inode number as a workaround for missing delete
   events on the Linux kernel

== 4.0: February 02, 2018

 - Warn instead of error if kqueue fails to register on STDIN
 - Close STDIN before running the utility when the restart option is used
 - Restore terminal settings if terminated by a signal

== 3.9: September 19, 2017

 - Fix use of poll(2) to avoid possible busy-loop on Linux
 - Disable keyboard input if reading STDIN fails

== 3.8: August 11, 2017

 - Run the utility if the spacebar is pressed
 - 'q' for quit

== 3.7: February 27, 2017


entr.c
/*
 * entr.c
 * A utility for running arbitrary commands when files change
 */

#include <sys/param.h>
#include <sys/resource.h>
#include <sys/stat.h>
#include <sys/sysctl.h>
#include <sys/wait.h>

#include <sys/event.h>

#include <dirent.h>
#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <libgen.h>
#include <limits.h>
#include <paths.h>
#include <signal.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <termios.h>
#include <time.h>
#include <unistd.h>

#include "missing/compat.h"

#include "data.h"
#include "status.h"

/* events to watch for */

#define NOTE_ALL NOTE_DELETE | NOTE_WRITE | NOTE_RENAME | NOTE_TRUNCATE | NOTE_ATTRIB

/* shared state */

extern int optind;
pid_t status_pid;
WatchFile **files;

/* globals */

WatchFile *leading_edge;
int child_pid;
int child_status;
int terminating;
int restart_signal;

int aggressive_opt;
int clear_opt;
int dirwatch_opt;
int noninteractive_opt;
int oneshot_opt;
int postpone_opt;
int restart_opt;
int shell_opt;
int status_filter_opt;

int termios_set;
struct termios canonical_tty;

static char *shell, *shell_base;
static char *argv0, *argv0_base;

/* function pointers */

int (*xstat)(const char *path, struct stat *sb);

/* forwards */

static void usage(bool);
static void terminate_utility();
static void set_restart_signal();
static void handle_exit(int sig);
static void proc_exit(int sig);
static void print_child_status(int status);
static int process_input(FILE *, WatchFile *[], int);
static int set_options(char *[]);
static int list_dir(char *);
static void run_utility(char *[]);
static void watch_file(int, WatchFile *);
static int compare_dir_contents(WatchFile *);
static void watch_loop(int, char *[]);

/*
 * The Event Notify Test Runner
 * run arbitrary commands when files change
 */
int
main(int argc, char *argv[]) {
	int kq;
	struct sigaction act;
	int ttyfd;
	short argv_index;
	int n_files;
	int i;
	struct kevent evSet;
	int open_max;

	/* call usage() if no command is supplied */
	if (argc < 2)
		usage(false);
	argv_index = set_options(argv);

	sigemptyset(&act.sa_mask);

	/* normally a user will exit this utility by do_execting Ctrl-C */
	act.sa_flags = SA_RESETHAND;
	act.sa_handler = handle_exit;
	if (sigaction(SIGINT, &act, NULL) != 0)
		err(1, "Failed to set SIGINT handler");
	if (sigaction(SIGTERM, &act, NULL) != 0)
		err(1, "Failed to set SIGTERM handler");
	if (sigaction(SIGHUP, &act, NULL) != 0)
		err(1, "Failed to set SIGHUP handler");

	set_restart_signal();

	/* notification used to combine the one-shot and restart options */
	act.sa_flags = 0;
	act.sa_handler = proc_exit;
	if (sigaction(SIGCHLD, &act, NULL) != 0)
		err(1, "Failed to set SIGCHLD handler");

	/* monitor symlinks if possible */
	xstat = stat;
#if defined(O_PATH) || defined(O_SYMLINK)
	if (getenv("ENTR_FOLLOW_SYMLINK") == NULL)
		xstat = lstat;
#endif

#if defined(_LINUX_PORT)
	/* attempt to read inotify limits */
	open_max = (unsigned) fs_sysctl(INOTIFY_MAX_USER_WATCHES);
	if (open_max == 0)
		open_max = 65536;
#elif defined(_MACOS_PORT)
	struct rlimit rl;
	int mib[2] = { CTL_KERN, KERN_MAXFILESPERPROC };
	size_t namelen = sizeof(open_max);

	if (getrlimit(RLIMIT_NOFILE, &rl) == -1)
		err(1, "getrlimit");
	if (sysctl(mib, 2, &open_max, &namelen, NULL, 0) == -1)
		open_max = OPEN_MAX;
	rl.rlim_cur = open_max;
	if (setrlimit(RLIMIT_NOFILE, &rl) != 0)
		err(1, "setrlimit cannot set rlim_cur to %u", open_max);
#else /* BSD */
	struct rlimit rl;

	if (getrlimit(RLIMIT_NOFILE, &rl) == -1)
		err(1, "getrlimit");
	open_max = (unsigned) rl.rlim_max;
	rl.rlim_cur = (rlim_t) open_max;
	if (setrlimit(RLIMIT_NOFILE, &rl) != 0)
		err(1, "setrlimit cannot set rlim_cur to %u", open_max);
#endif

	if (getenv("EV_TRACE"))
		fprintf(stderr, "open_max: %d\n", open_max);

	/* prevent interactive utilities from paging output */
	setenv("PAGER", "/bin/cat", 0);

	/* ensure a shell is available to use */
	if ((shell = getenv("SHELL")) == NULL)
		shell = "/bin/sh";
	shell_base = strdup(shell);
	shell_base = basename(shell_base);

	/* initialize status filter */
	if (shell_opt)
		argv0 = shell;
	else
		argv0 = (argv + argv_index)[0];
	argv0_base = basename(argv0);
	if (status_filter_opt)
		start_log_filter(status_filter_opt);

	/* drop privileges */
	if (pledge("stdio rpath tty proc exec", NULL) == -1)
		err(1, "pledge");

	/* sequential scan may depend on a 0 at the end */
	files = calloc(open_max + 1, sizeof(WatchFile *));
	if (files == NULL)
		err(1, "calloc");

	if ((kq = kqueue()) == -1)
		err(1, "cannot create kqueue");

	/* expect file list from a pipe */
	if (isatty(fileno(stdin)))
		usage(false);

	/* read input and populate watch list, skipping non-regular files */
	n_files = process_input(stdin, files, open_max);
	if (n_files == 0)
		errx(1, "No regular files to watch");
	if (n_files == -1)
		errx(1,
		    "Too many files listed; the hard limit for your login"
		    " class is %u. Please consult"
		    " http://eradman.com/entrproject/limits.html",
		    open_max);
	for (i = 0; i < n_files; i++)
		watch_file(kq, files[i]);

	if (!noninteractive_opt) {
		/* Attempt to open a tty so that editors don't complain */
		ttyfd = open(_PATH_TTY, O_RDONLY);
		if (ttyfd > STDIN_FILENO) {
			if (dup2(ttyfd, STDIN_FILENO) != 0)
				warnx("can't dup2 to stdin");
			close(ttyfd);
		}

		/* remember terminal settings */
		if (tcgetattr(STDIN_FILENO, &canonical_tty) == -1)
			errx(1, "unable to get terminal attributes, use '-n' to run non-interactively");

		/* Use keyboard input as a trigger */
		EV_SET(&evSet, STDIN_FILENO, EVFILT_READ, EV_ADD, NOTE_LOWAT, 1, NULL);
		if (kevent(kq, &evSet, 1, NULL, 0, NULL) == -1)
			warnx("failed to register stdin");
	}

	watch_loop(kq, argv + argv_index);
	return 1;
}

/* Utility functions */

void
usage(bool summary) {
	fprintf(stderr, "release: %s\n", RELEASE);
	fprintf(stderr, "usage: entr [-acdnprsxz] utility [argument [/_] ...] < filenames\n");
	if (!summary) {
		fprintf(stderr, "hint: use -h to display option summary\n");
		goto end;
	}

	printf("summary:\n"
	       "    -a  Do not consolidate events\n"
	       "    -c  Clear screen before execution\n"
	       "    -d  Track files added or removed from directories\n"
	       "    -n  Non-interactive mode\n"
	       "    -p  Wait for first event\n"
	       "    -r  Run as a background process, use signal to restart\n"
	       "    -s  Evaluate using a shell\n"
	       "    -x  Format exit status\n"
	       "    -z  Exit after the utility completes\n");
	printf("docs:\n"
	       "    man entr\n");

end:
	exit(1);
}

void
terminate_utility() {
	int status;

	terminating = 1;

	if (child_pid > 0) {
		killpg(child_pid, restart_signal);
		waitpid(child_pid, &status, 0);
		child_pid = 0;
	}

	terminating = 0;
}

void
set_restart_signal() {
	const char *sig;
	const int signum[] = { SIGHUP, SIGINT, SIGQUIT, SIGTERM, SIGUSR1, SIGUSR2 };
	const char *signame[] = { "HUP", "INT", "QUIT", "TERM", "USR1", "USR2" };
	int i;

	if ((sig = getenv("ENTR_RESTART_SIGNAL")) == NULL) {
		restart_signal = SIGTERM;
		return;
	}

	if (strncmp(sig, "SIG", 3) == 0)
		sig += 3;

	for (i = 0; signum[i] < 6; i++) {
		if (strcmp(sig, signame[i]) == 0)
			restart_signal = signum[i];
	}

	if (restart_signal == 0)
		errx(1, "unrecognized signal: %s <> (HUP, INT, QUIT, TERM, USR1, USR2)", sig);
}

/* Callbacks */

void
handle_exit(int sig) {
	if ((!noninteractive_opt) && (termios_set))
		tcsetattr(STDIN_FILENO, TCSADRAIN, &canonical_tty);

	terminate_utility();

	if (status_filter_opt)
		end_log_filter();

	if ((sig == SIGINT || sig == SIGHUP))
		_exit(0);
	else
		raise(sig);
}

void
proc_exit(int sig) {
	int status;
	int saved_errno = errno;

	if (status_filter_opt && (terminating == 0)) {
		if (waitpid(status_pid, &status, WNOHANG) > 0) {
			if (WIFSIGNALED(status)) {
				terminating = 1;
				warnx("status process killed by signal");
				kill(getpid(), SIGINT);
			}
			if (WIFEXITED(status)) {
				terminating = 1;
				warnx("status process terminated");
				kill(getpid(), SIGINT);
			}
		}
	}

	if (waitpid(child_pid, &status, 0) != -1) {
		child_status = status;

		if ((!noninteractive_opt) && (termios_set))
			tcsetattr(STDIN_FILENO, TCSADRAIN, &canonical_tty);

		if ((oneshot_opt == 1) && (terminating == 0)) {
			if (restart_opt == 0)
				print_child_status(child_status);

			if (WIFSIGNALED(child_status))
				_exit(128 + WTERMSIG(child_status));
			else
				_exit(WEXITSTATUS(child_status));
		}
	}
	/* restore errno so that the resuming code is unimpacted. */
	errno = saved_errno;
}

void
print_child_status(int status) {
	int len;
	char buf[2048];

	if (status_filter_opt) {
		if (WIFSIGNALED(status))
			len = snprintf(buf, sizeof(buf), "signal|%d|%s\n", WTERMSIG(status), argv0_base);
		else
			len = snprintf(buf, sizeof(buf), "exit|%d|%s\n", WEXITSTATUS(status), argv0_base);
		write_log_filter(buf, len);
	}
}

/*
 * Read lines from a file stream (normally STDIN).  Returns the number of
 * regular files to be watched or -1 if max_files is exceeded.
 */
int
process_input(FILE *file, WatchFile *files[], int max_files) {
	char buf[PATH_MAX + 1];
	char *p, *path, *parent_path;
	int n_files = 0;
	struct stat sb;
	int i, matches;
	size_t len;

	while (fgets(buf, sizeof(buf), file) != NULL) {
		if ((p = strchr(buf, '\n')) != NULL)
			*p = '\0';
		if (buf[0] == '\0')
			continue;
		path = &buf[0];

		len = strlen(path);
		if ((len + 1) > PATH_MAX) {
			path[100] = '\0';
			errx(1, "path too long: %s..", path);
		}

		if (xstat(path, &sb) == -1) {
			warnx("unable to stat '%s'", path);
			continue;
		}

		if ((S_ISREG(sb.st_mode) | S_ISLNK(sb.st_mode)) != 0) {
			files[n_files] = malloc(sizeof(WatchFile));
			if (files[n_files] == NULL)
				err(1, "malloc");
			memcpy(files[n_files]->fn, path, len + 1);
			files[n_files]->is_dir = 0;
			files[n_files]->is_symlink = (S_ISLNK(sb.st_mode) != 0) ? 1 : 0;
			files[n_files]->file_count = 0;
			files[n_files]->mode = sb.st_mode;
			files[n_files]->ino = sb.st_ino;
			n_files++;

			/* also watch the directory if it's not already in the list */
			if (dirwatch_opt > 0) {
				if ((parent_path = dirname(path)) == 0)
					err(1, "dirname '%s' failed", path);
				for (matches = 0, i = 0; i < n_files; i++) {
					if ((files[i]->is_dir == 1) && (strcmp(files[i]->fn, parent_path) == 0))
						matches++;
				}
				if (matches == 0) {
					if (stat(parent_path, &sb) == -1)
						warnx("unable to stat '%s'", parent_path);
					path = parent_path;
				}
			}
		}
		if (S_ISDIR(sb.st_mode) != 0) {
			files[n_files] = malloc(sizeof(WatchFile));
			if (files[n_files] == NULL)
				err(1, "malloc");
			memcpy(files[n_files]->fn, path, len + 1);
			files[n_files]->is_dir = 1;
			files[n_files]->is_symlink = 0;
			files[n_files]->file_count = list_dir(path);
			files[n_files]->mode = sb.st_mode;
			files[n_files]->ino = sb.st_ino;
			n_files++;
		}
		if (n_files + 1 > max_files)
			return -1;
	}
	return n_files;
}

int
list_dir(char *dir) {
	struct dirent *dp;
	DIR *dfd = opendir(dir);
	int count = 0;

	if (dfd == NULL)
		errx(1, "unable to open directory: '%s'", dir);
	while ((dp = readdir(dfd)) != NULL)
		if ((dirwatch_opt == 2) || (dp->d_name[0] != '.'))
			count++;
	closedir(dfd);
	return count;
}

/*
 * Evaluate command line arguments and return an offset to the command to
 * execute.
 */
int
set_options(char *argv[]) {
	int ch;
	int argc;

	if (argv[1] && strcmp(argv[1], "-h") == 0)
		usage(true);

	/* read arguments until we reach a command */
	for (argc = 1; argv[argc] != 0 && argv[argc][0] == '-'; argc++)
		;
	while ((ch = getopt(argc, argv, "acdnprsxz")) != -1) {
		switch (ch) {
		case 'a':
			aggressive_opt = 1;
			break;
		case 'c':
			clear_opt = clear_opt ? 2 : 1;
			break;
		case 'd':
			dirwatch_opt = dirwatch_opt ? 2 : 1;
			break;
		case 'n':
			noninteractive_opt = 1;
			break;
		case 'p':
			postpone_opt = 1;
			break;
		case 'r':
			restart_opt = 1;
			break;
		case 's':
			shell_opt = 1;
			break;
		case 'x':
			status_filter_opt = status_filter_opt ? 2 : 1;
			break;
		case 'z':
			oneshot_opt = 1;
			break;
		default:
			usage(false);
		}
	}
	if (argv[optind] == 0)
		usage(false);

	if (status_filter_opt && restart_opt)
		errx(1, "-r and -x may not be combined");

	if ((shell_opt == 1) && (argv[optind + 1] != 0))
		errx(1, "-s requires commands to be formatted as a single argument");
	return optind;
}

/*
 * Execute the program supplied on the command line. If restart was set
 * then send the child process SIGTERM and restart it.
 */
void
run_utility(char *argv[]) {
	int pid;
	int i, m;
	int ret, status;
	struct timespec delay = { 0, 1000000 };
	char **new_argv;
	char *p, *arg_buf, *src;
	int argc;
	size_t len, rem;

	if (restart_opt == 1) {
		terminate_utility();
		nanosleep(&delay, NULL);
	}

	arg_buf = malloc(rem = ARG_MAX);
	if (arg_buf == NULL)
		err(1, "malloc");

	if (shell_opt == 1) {
		/* run argv[1] with a shell using the leading edge as $0 */
		argc = 4;
		new_argv = calloc(argc + 1, sizeof(char *));
		if (new_argv == NULL)
			err(1, "calloc");
		new_argv[0] = shell;
		new_argv[1] = "-c";
		new_argv[2] = argv[0];
		new_argv[3] = leading_edge->fn;
	} else {
		/* clone argv on each invocation to allow for substitution rules */
		for (argc = 0; argv[argc]; argc++)
			;
		new_argv = calloc(argc + 1, sizeof(char *));
		if (new_argv == NULL)
			err(1, "calloc");
		new_argv[0] = "/bin/false";
		for (m = 0, i = 0, p = arg_buf; i < argc; i++) {
			new_argv[i] = p;
			if ((m < 1) && (strcmp(argv[i], "/_")) == 0) {
				src = leading_edge->fn;
				m++;
			} else
				src = argv[i];

			len = strlen(src);
			if (len + 1 > rem)
				errx(1, "argument list too long");

			memcpy(p, src, len + 1);
			p += len + 1;
			rem -= len + 1;
		}
	}

	pid = fork();
	if (pid == -1)
		err(1, "can't fork");

	if (pid == 0) {
		/* 2J - erase the entire display
		 * 3J - clear scrollback buffer
		 * H  - set cursor position to the default
		 */
		if (clear_opt == 1)
			printf("\033[2J\033[H");
		if (clear_opt == 2)
			printf("\033[2J\033[3J\033[H");
		fflush(stdout);

		/* Set process group so subprocess can be signaled */
		if (restart_opt == 1) {
			setpgid(0, getpid());
			close(STDIN_FILENO);
			open(_PATH_DEVNULL, O_RDONLY);
		}
		/* wait up to 1 seconds for each file to become available */
		for (i = 0; i < 10; i++) {
			ret = execvp(new_argv[0], new_argv);
			if (errno == ETXTBSY)
				nanosleep(&delay, NULL);
			else
				break;
		}
		if (ret != 0)
			err(1, "exec %s", new_argv[0]);
	}
	child_pid = pid;

	if (restart_opt == 0 && oneshot_opt == 0) {
		if (waitpid(child_pid, &status, 0) != -1)
			child_status = status;

		print_child_status(child_status);
	}

	free(arg_buf);
	free(new_argv);
}

/*
 * Wait for file to become accessible and register a kevent to watch it
 */
void
watch_file(int kq, WatchFile *file) {
	struct kevent evSet;
	int i = 0;
	struct timespec delay = { 0, 100 * 1000000 };

	/* wait up to 1 second for file to become available */
	for (;;) {
#if defined(O_EVTONLY)
		file->fd = open(file->fn, O_RDONLY | O_CLOEXEC | O_EVTONLY | O_SYMLINK);
#elif defined(O_PATH)
		file->fd = open(file->fn, O_RDONLY | O_CLOEXEC | O_PATH | O_NOFOLLOW);
#else
		file->fd = open(file->fn, O_RDONLY | O_CLOEXEC);
#endif
		if (file->fd == -1) {
			if (i < 10)
				nanosleep(&delay, NULL);
			else {
				warn("cannot open '%s'", file->fn);
				terminate_utility();
				exit(1);
			}
		} else
			break;
		i++;
	}

	EV_SET(&evSet, file->fd, EVFILT_VNODE, EV_ADD | EV_CLEAR, NOTE_ALL, 0, file);
	if (kevent(kq, &evSet, 1, NULL, 0, NULL) == -1) {
		if (errno == ENOSPC)
			errx(1,
			    "Unable to allocate memory for kernel queue."
			    " Please consult"
			    " http://eradman.com/entrproject/limits.html");
		else
			err(1, "failed to register VNODE event");
	}
}

/*
 * Wait for directory contents to stabilize
 */
int
compare_dir_contents(WatchFile *file) {
	int i;
	struct timespec delay = { 0, 100 * 1000000 };

	/* wait up to 0.5 seconds for file to become available */
	for (i = 0; i < 5; i++) {
		if (list_dir(file->fn) == file->file_count)
			return 0;
		nanosleep(&delay, NULL);
	}
	return 1;
}

/*
 * Wait for events to and execute a command. Four major concerns are in play:
 *   leading_edge: Global reference to the first file to have changed
 *   reopen_only : Unlink or rename events which require us to spin while
 *                 waiting for the file to reappear. These must always be
 *                 processed
 *   collate_only: Changes that indicate that more events are likely to occur.
 *                 Watch for more events using a short timeout
 *   do_exec     : Delay execution until all events have been processed. Allow
 *                 the user to edit files while the utility is running without
 *                 any visible side-effects
 *   dir_modified: The number of files changed for a directory under watch
 */
void
watch_loop(int kq, char *argv[]) {
	struct kevent evSet;
	struct kevent evList[32];
	int nev;
	WatchFile *file;
	int i;
	struct timespec evTimeout = { 0, 1000000 };
	int reopen_only = !aggressive_opt;
	int collate_only = 0;
	int do_exec = 0;
	int dir_modified = 0;
	int leading_edge_set = 0;
	struct stat sb;
	char c;
	struct termios character_tty;

	leading_edge = files[0]; /* default */
	if (postpone_opt == 0)
		run_utility(argv);

	if (!noninteractive_opt) {
		/* disabling/restore line buffering and local echo */
		character_tty = canonical_tty;
		character_tty.c_lflag &= ~(ICANON | ECHO);
	}

main:
	if (!noninteractive_opt) {
		tcsetattr(STDIN_FILENO, TCSADRAIN, &character_tty);
		termios_set = 1;
	}

	if ((reopen_only == 1) || (collate_only == 1)) {
		nev = kevent(kq, NULL, 0, evList, 32, &evTimeout);
	} else {
		nev = kevent(kq, NULL, 0, evList, 32, NULL);
		dir_modified = 0;
	}

	if ((nev == -1) && (errno != EINTR))
		warn("kevent failed");

	for (i = 0; i < nev; i++) {
		if (!noninteractive_opt && evList[i].filter == EVFILT_READ) {
			if (read(STDIN_FILENO, &c, 1) < 1) {
				EV_SET(&evSet, STDIN_FILENO, EVFILT_READ, EV_DELETE, NOTE_LOWAT, 0, NULL);
				if (kevent(kq, &evSet, 1, NULL, 0, NULL) == -1)
					err(1, "failed to remove READ event");
			} else {
				if (c == ' ')
					do_exec = 1;
				if (c == 'q')
					kill(getpid(), SIGINT);
			}
		}
		if (evList[i].filter != EVFILT_VNODE)
			continue;

		file = (WatchFile *) evList[i].udata;
		if (file->is_dir == 1)
			dir_modified += compare_dir_contents(file);
	}
	if (!noninteractive_opt)
		tcsetattr(STDIN_FILENO, TCSADRAIN, &canonical_tty);

	collate_only = 0;
	for (i = 0; i < nev; i++) {
		if (evList[i].filter != EVFILT_VNODE)
			continue;
		file = (WatchFile *) evList[i].udata;
		if (evList[i].fflags & NOTE_DELETE || evList[i].fflags & NOTE_RENAME) {
			EV_SET(&evSet, file->fd, EVFILT_VNODE, EV_DELETE, NOTE_ALL, 0, file);
			if (kevent(kq, &evSet, 1, NULL, 0, NULL) == -1)
				err(1, "failed to remove VNODE event");
#if !defined(_LINUX_PORT)
			/* free file descriptor no longer monitored by kqueue */
			if ((file->fd != -1) && (close(file->fd) == -1))
				err(1, "unable to close file");
#endif
			watch_file(kq, file);
			collate_only = 1;
		}
	}
	if (reopen_only == 1) {
		reopen_only = 0;
		goto main;
	}

	for (i = 0; i < nev && reopen_only == 0; i++) {

		if (evList[i].filter != EVFILT_VNODE)
			continue;
		file = (WatchFile *) evList[i].udata;
		if ((file->is_dir == 1) && (dir_modified == 0))
			continue;

		if (evList[i].fflags & NOTE_DELETE || evList[i].fflags & NOTE_WRITE
		    || evList[i].fflags & NOTE_RENAME || evList[i].fflags & NOTE_TRUNCATE) {
			if ((dir_modified > 0) && (restart_opt == 1))
				continue;
			do_exec = 1;
		}

		if (evList[i].fflags & NOTE_ATTRIB && S_ISREG(file->mode) != 0
		    && xstat(file->fn, &sb) == 0) {
			if (file->mode != sb.st_mode) {
				do_exec = 1;
				file->mode = sb.st_mode;
			}
			if (file->ino != sb.st_ino) {
#if defined(_LINUX_PORT)
				do_exec = 1;
#endif
				file->ino = sb.st_ino;
			}
		} else if (evList[i].fflags & NOTE_ATTRIB)
			continue;

		if ((leading_edge_set == 0) && (file->is_dir == 0) && (do_exec == 1)) {
			leading_edge = file;
			leading_edge_set = 1;
		}

		if (getenv("EV_TRACE")) {
			fprintf(stderr, "%d/%d: fflags: 0x%x %s %o %s\n", i, nev, evList[i].fflags,
			    file->is_dir ? "d" : "r", file->mode, file->fn);
		}
	}

	if (collate_only == 1)
		goto main;
	if (do_exec == 1) {
		do_exec = 0;
		run_utility(argv);
		if (!aggressive_opt)
			reopen_only = 1;
		leading_edge_set = 0;
	}
	if (dir_modified > 0) {
		terminate_utility();
		errx(2, "directory altered");
	}

	goto main;
}
